src/webgpu/shader/validation/uniformity/uniformity.spec.ts

```diff
@@ -0,0 +1,277 @@
+  { op: 'dpdyFine', stage: 'fragment' },
+  { op: 'fwidth', stage: 'fragment' },
+  { op: 'fwidthCoarse', stage: 'fragment' },
+  { op: 'fwidthFine', stage: 'fragment' },
+  { op: 'storageBarrier', stage: 'compute' },
+  { op: 'workgroupBarrier', stage: 'compute' },
+  { op: 'workgroupUniformLoad', stage: 'compute' },
+] as const;
+
+export const kConditions = [
+  { cond: 'uniform_storage_ro', expectation: true },
+  { cond: 'nonuniform_storage_ro', expectation: false },
+  { cond: 'nonuniform_storage_rw', expectation: false },
+  { cond: 'nonuniform_builtin', expectation: false },
+  { cond: 'uniform_literal', expectation: true },
+  { cond: 'uniform_const', expectation: true },
+  { cond: 'uniform_override', expectation: true },
+  { cond: 'uniform_let', expectation: true },
+  { cond: 'nonuniform_let', expectation: false },
+  { cond: 'uniform_or', expectation: true },
+  { cond: 'nonuniform_or1', expectation: false },
+  { cond: 'nonuniform_or2', expectation: false },
+  { cond: 'uniform_and', expectation: true },
+  { cond: 'nonuniform_and1', expectation: false },
+  { cond: 'nonuniform_and2', expectation: false },
+  { cond: 'uniform_func_var', expectation: true },
+  { cond: 'nonuniform_func_var', expectation: false },
+] as const;
+
+export function generateCondition({ condition }: { condition: string }) {
+  let code = ``;
+
+  switch (condition) {
+    case 'uniform_storage_ro': {
+      code += `ro_buffer[0] == 0`;
+      break;
+    }
+    case 'nonuniform_storage_ro': {
+      code += `ro_buffer[priv_var[0]] == 0`;
+      break;
+    }
+    case 'nonuniform_storage_rw': {
+      code += `rw_buffer[0] == 0`;
+      break;
+    }
+    case 'nonuniform_builtin': {
+      code += `p.x == 0`;
+      break;
+    }
+    case 'uniform_literal': {
+      code += `false`;
+      break;
+    }
+    case 'uniform_const': {
+      code += `c`;
+      break;
+    }
+    case 'uniform_override': {
+      code += `o == 0`;
+      break;
+    }
+    case 'uniform_let': {
+      code += `u_let == 0`;
+      break;
+    }
+    case 'nonuniform_let': {
+      code += `n_let == 0`;
+      break;
+    }
+    case 'uniform_or': {
+      code += `u_let == 0 || uniform_buffer.y > 1`;
+      break;
+    }
+    case 'nonuniform_or1': {
+      code += `u_let == 0 || n_let == 0`;
+      break;
+    }
+    case 'nonuniform_or2': {
+      code += `n_let == 0 || u_let == 0`;
+      break;
+    }
+    case 'uniform_and': {
+      code += `u_let == 0 && uniform_buffer.y > 1`;
+      break;
+    }
+    case 'nonuniform_and1': {
+      code += `u_let == 0 && n_let == 0`;
+      break;
+    }
+    case 'nonuniform_and2': {
+      code += `n_let == 0 && u_let == 0`;
+      break;
+    }
+    case 'uniform_func_var': {
+      code += `u_f == 0`;
+      break;
+    }
+    case 'nonuniform_func_var': {
+      code += `n_f == 0`;
+      break;
+    }
+    default: {
+      break;
+    }
+  }
+
+  return code;
+}
+
+export function generateOp({ op }: { op: string }) {
```

Comment:
As with `generateConditionalStatement` below, I am confused by the use of structured input here, and also I think a return type should be specified.

```
function generateOp(op: string) : string
```